server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -660,31 +711,107 @@ private static int createByteVector(final FlatBufferBuilder builder, final byte[
         return builder.endVector();
     }
 
-    private long appendAddColumns(final View view, final long startRange, final long endRange,
-            final Consumer<InputStream> addStream,
-            final ChunkInputStreamGenerator.FieldNodeListener fieldNodeListener,
-            final ChunkInputStreamGenerator.BufferListener bufferListener) throws IOException {
+    private void processBatches(Consumer<InputStream> visitor, final View view,
+            final long numRows, final int maxBatchSize, ByteBuffer metadata,
+            final ColumnVisitor columnVisitor, final MutableLong bytesWritten) throws IOException {
+        long offset = 0;
+        MutableInt actualBatchSize = new MutableInt();
 
+        int batchSize = DEFAULT_INITIAL_BATCH_SIZE;
 
-        try (final RowSet myAddedOffsets = view.addRowOffsets().subSetByPositionRange(startRange, endRange)) {
-            // add the add-column streams
-            for (final ChunkInputStreamGenerator col : addColumnData) {
-                final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
-                        col.getInputStream(view.options(), myAddedOffsets);
-                drainableColumn.visitFieldNodes(fieldNodeListener);
-                drainableColumn.visitBuffers(bufferListener);
+        while (offset < numRows) {
+            try {
+                final InputStream is = getInputStream(
+                        view, offset, batchSize, actualBatchSize, metadata, columnVisitor);
+                int bytesToWrite = is.available();
+
+                // treat this as a hard limit, exceeding fails a client or w2w (unless we are sending a single
+                // row then we must send and let it potentially fail)
+                if (bytesToWrite < DEFAULT_MESSAGE_SIZE_LIMIT || batchSize == 1) {
+                    // let's write the data
+                    visitor.accept(is);
+
+                    bytesWritten.add(bytesToWrite);
+                    offset += actualBatchSize.intValue();
+                    metadata = null;
+                } else {
+                    // can't write this, so close the input stream and retry
+                    is.close();
+                }
+                // recompute the batch limit for the next message
+                int bytesPerRow = bytesToWrite / actualBatchSize.intValue();
+                if (bytesPerRow > 0) {
+                    int rowLimit = DEFAULT_MESSAGE_SIZE_LIMIT / bytesPerRow;
 
-                // Add the drainable last as it is allowed to immediately close a row set the visitors need
-                addStream.accept(drainableColumn);
+                    // add some margin for abnormal cell contents
+                    batchSize = Math.min(maxBatchSize, Math.max(1, (int) ((double) rowLimit * 0.9)));
+                }
+            } catch (SizeException ex) {
+                // was an overflow in the ChunkInputStream generator (probably VarBinary). We can't compute the
```

Comment:
You need to handle the case when `batchSize` is already 1; at least for completeness if it is currently impossible. Though I think it is possible if `addStream.accept` can throw a SizeException for the sum of columns exceeding max int.